Nullable.md

```diff
@@ -0,0 +1,442 @@
+we are expected to evaluate `f` in that scenario. It is only a problem when we
+need to know the return type of `f`, but must not determine this type by
+any actual evaluation of `f`. Because this problem does not occur in the
+non-null case, it does arise for either `Union{T, Void}` or
+`Union{Some{T}, Void}`. The problem is unique to `Union{T, Null{T}}` and our
+current `Nullable{T}` implementation as one must determine the type parameter
+`T` even for the null case.
+
+## Step 2:: Define Core Operations on Nullable Types by Lifting
+
+Once an implementation of `Nullable{T}` is agreed upon, the core semantic issue
+for nullable types is maximizing code reuse by making lifting easy to use. In
+other words, we want to make it easy to evaluate `f(x::Nullable{T})` without
+requiring users to define methods for every function to handle the null case,
+for which the semantics are so consistent. We want to avoid requiring that
+users write code for the null case because we know that the overwhelming
+majority of functions will return `null` as their output whenever any of
+their arguments are `null`.
+
+As mentioned earlier, extending the definition of functions from `T` to `?T`
+is called lifting. We propose the following extensible approach to implementing
+lifting that defaults to the semantics we expect will occur most commonly, but
+allows implementing alternative semantics where needed.
+
+First, we define `broadcast(f, x::Nullable)` as follows:
+
+```
+immutable Lifted{T}
+    func::T
+end
+
+lift(f) = Lifted(f)
+
+broadcast(f, x::Nullable) = broadcast(lift(f), x)
+
+broadcast(f′::Lifted, x::Nullable) = hasvalue(x) ? f′.func(value(x)) : null
+```
+
+This allows us to give default semantics to all functions that implement the
+natural approach to lifting, while still making it possible to handle special
+case functions that require distinct semantics.
+
+Some such functions are:
+
+* `isnull(x::?T)`
+* `get(x::?T)`
+* `get(x::?T, default::T)`
+
+Because we would use broadcasting, lifting would happen via dot-syntax:
+for example, `x .+ y` would be lifted by default. In addition, we would special
+several operators like `+`, `-`, `==`, etc. to automatically be equivalent to
+broadcasting on nullables.
+
+## Step 3: Introduce Syntactic Sugar for Nullable Types
+
+Following the lead of
+[Flow for Javascript](https://flowtype.org/docs/nullable-types.html), we
+would introduce `?T` as sugar for `Nullable{T}`. We have also considered C#'s
+`T?`. Which is chosen will likely depend upon how they interact with the
+ternary operator.
+
+We might also introduce a null-coalescing operator, `??`, with
+right-associativity that essentially performs lifting: for example,
+`x ?? y ?? 0` would be equivalent to `hasvalue(x) ? x : (hasvalue(y) ? y : 0)`.
+
+Finally, we might introduce `.?` for lifted field access such that
+`x?.field_name` would be equivalent to `getfield.(x, :field_name)`.
```

Comment:
You first mention `.?` but then use `x?.`. Which one is being considered? Or are both?